New configuration format.

xpconanfan opened the pull request:
Fixes #42 

* Introducing new YAML format.
* Deprecating the old JSON format.
* Separate controller info and test params.
* No more implicit value inheritance, use YAML's anchor/reference instead.
* LogPath now has default value, this allows executing tests without config.
* Update README.
* Update unit tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/mobly/85)
<!-- Reviewable:end -->

dthkao commented on README.md:
````diff
@@ -72,20 +72,14 @@ restrictions.
+**sample_config.yml**
+
+```yaml
+TestBeds:
````
Are we going with testbed as one word or two? If one perhaps we should lowercase all the B's in labels and names (for consistency)

xpconanfan replied:
This is two words on purpose here.

dthkao replied:
Then can we make the comments consistent?

xpconanfan replied:
Done

dthkao commented on README.md:
````diff
@@ -113,21 +107,22 @@ if __name__ == "__main__":
-]
+```yaml
+AndroidDevice:
+  - serial: xyz,
````
No need to change anything, but github's code review markers make yaml confusing to review..

dthkao commented on README.md:
````diff
@@ -211,33 +201,37 @@ In the test script, you could access the user parameter:
 ```
 
-## Example 4: Multiple Test Beds
+## Example 4: Multiple Test Beds and Default Test Params
````
Full word (Parameters)?

xpconanfan replied:
Done

xpconanfan commented:
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/google/mobly/85#-Kc2S7YdpoWriO8Wa6wS:-Kc2S7YeGTiVBIQF21sR:b1f4yu5) (no related file):*
Testing out reviewable with this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Ping?

adorokhine commented:
Reviewed 6 of 9 files at r1, 1 of 1 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/google/mobly/85#-KcVLsTW4KC_CEGYcrGm:-KcVLsTXophA4u52iaxk:becmd52) (no related file):*
android_device has some references to json configs in comments

---

*[README.md, line 213 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGMsh-jB4DfpfrUFL:-KcVGMsii5KLCg0QSlVV:b-d0ja83) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/README.md#L213)):*
> ```Markdown
> 
> ```yaml
> DefaultParams: &DefaultParams
> ```

Might want to add a comment that DefaultParams is optional and clarify what it does for the noobs:
"Optional: create an anchor for a set of parameters shared between all testbeds"

---

*[setup.py, line 28 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGqMKvoR1czWx45KQ:-KcVGqMKvoR1czWx45KR:bf4nz99) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/setup.py#L28)):*
> ```Python
>     # which is what comes with Ubuntu 14.04 LTS.
>     'mock<=1.0.1',
>     'pyyaml'
> ```

Mind realphabetizing this list?

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-KcVH2Vu3mYM_3igRdc2:bycvybx) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
> ```Markdown
> **sample_config.yml**
> 
> ```yaml
> ```

Is this just a duplicate of the readme? Can we delete one copy?

---

*[docs/tutorials/lesson1.md, line 48 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVHSMQRz0GEADYQW0c:-KcVHSMRsu5ZSwIJWOGJ:bay6vei) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L48)):*
> ```Markdown
> *To execute:*
> 
>     $ python hello_world_test.py -c sample_config.json
> ```

json? (here and elsewhere)

---

*[mobly/config_parser.py, line 69 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVIOxOLQygBqDiZJpY:-KcVIOxOLQygBqDiZJpZ:bovq7d8) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L69)):*
> ```Python
> def _validate_testbed_configs(testbed_configs):
>     """Validates the testbed configurations.
> ```

l is not defined in this scope

---

*[mobly/config_parser.py, line 192 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVIq5T1LD3BE41eWAJ:-KcVIq5U214pvjmAxEGV:b-6cvfw5) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L192)):*
> ```Python
>         log_path = os.environ[_ENV_MOBLY_LOGPATH]
>     log_path = utils.abs_path(log_path)
>     print(log_path)
> ```

Should use logging here instead of print. Also needs some context, otherwise it's just a random path flying by.

---

*[mobly/config_parser.py, line 199 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJ2tYIsPhaqDsX9z4:-KcVJ2tZC3AixPRj8Cze:b-xn2ljy) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L199)):*
> ```Python
>     test_configs = []
>     for original_bed_config in configs[keys.Config.key_testbed.value]:
>         # Throw an error for users using old config format.
> ```

Not sure where in the code this happens? I don't see any errors or even [] dict lookups

---

*[mobly/config_parser.py, line 204 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJMIgIe5gYuXc1zWc:-KcVJMIgIe5gYuXc1zWd:b-mhs0e5) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L204)):*
> ```Python
>         test_params = original_bed_config.get(
>             keys.Config.key_testbed_test_params.value, {})
>         new_test_config = {
> ```

Since this object seems to have fixed keys and is part of an API, seems like it should be a real python object instead of a dict.

---

*[mobly/config_parser.py, line 205 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJBha59azoS3y41EZ:-KcVJBha59azoS3y41E_:b5n36nf) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L205)):*
> ```Python
>             keys.Config.key_testbed_test_params.value, {})
>         new_test_config = {
>             keys.Config.ikey_testbed_name.value:
> ```

Why do we need separate external and internal representations?

---

*[mobly/config_parser.py, line 237 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJh3J9zJALRghZpJG:-KcVJh3KCSNpvtDawnyH:b-122iod) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L237)):*
> ```Python
> 
> 
> def load_config_file(path):
> ```

Is this supposed to be private to the module?
If you don't want external callers, needs an underscore.

---

*[mobly/keys.py, line 30 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJoUu9AEfz6u0BlrX:-KcVJoUu9AEfz6u0BlrY:b-7lccw) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/keys.py#L30)):*
> ```Python
>     key_log_path = "LogPath"
>     # Top-level keywordss used in a test bed config.
>     key_testbed = "TestBeds"
> ```

based on the convention above, seems like this should be key_test_beds?

---

*[mobly/test_runner.py, line 201 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVK7N0CY5P2Ka1ja2J:-KcVK7N1uBwbaWyLooYe:b-a83qwy) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/test_runner.py#L201)):*
> ```Python
>         self.controller_configs = self.test_configs[
>             keys.Config.ikey_testbed_controllers.value]
>         self.testbed_name = self.test_configs[keys.Config.ikey_testbed_name.
> ```

Different newline conventions between this line and the line above?

---

*[tests/mobly/test_runner_test.py, line 44 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVKt-ryjP7ogDePwQ1:-KcVKt-s8kUoy0vG0Z_j:b-k2vw4) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/tests/mobly/test_runner_test.py#L44)):*
> ```Python
>     def setUp(self):
>         self.tmp_dir = tempfile.mkdtemp()
>         self.base_mock_test_config = {
> ```

Why is this config using internal keys? This used to look like a parsed external config

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[README.md, line 213 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGMsh-jB4DfpfrUFL:-KcVSRjYDgcboFsDxckz:bb6un1j) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/README.md#L213)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Might want to add a comment that DefaultParams is optional and clarify what it does for the noobs:
"Optional: create an anchor for a set of parameters shared between all testbeds"
</blockquote></details>

Done

---

*[setup.py, line 28 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGqMKvoR1czWx45KQ:-KcVSX_TpvzEtrdfIQQa:bb6un1j) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/setup.py#L28)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Mind realphabetizing this list?
</blockquote></details>

Done

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-KcVSbJenSgo8q6Og3SA:b-rgiy1g) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Is this just a duplicate of the readme? Can we delete one copy?
</blockquote></details>

All the docs files were copied from readme... do we want to remove them for good?
We never got the time to make them pretty... :(

---

*[docs/tutorials/lesson1.md, line 48 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVHSMQRz0GEADYQW0c:-KcVVWRl_uqKSbS6Wiem:bb6un1j) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L48)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

json? (here and elsewhere)
</blockquote></details>

Done

---

*[mobly/config_parser.py, line 69 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVIOxOLQygBqDiZJpY:-KcVT7Z9zt_bONYno64v:bb6un1j) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L69)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

l is not defined in this scope
</blockquote></details>

Done

---

*[mobly/config_parser.py, line 192 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVIq5T1LD3BE41eWAJ:-KcVTFCND6WHERhk2JWF:b-il2uiy) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L192)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Should use logging here instead of print. Also needs some context, otherwise it's just a random path flying by.
</blockquote></details>

Removed. this was leftover from debuggin

---

*[mobly/config_parser.py, line 199 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJ2tYIsPhaqDsX9z4:-KcVTcf1E7u2r85aL9qb:b1tge1s) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L199)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Not sure where in the code this happens? I don't see any errors or even [] dict lookups
</blockquote></details>

Removed.
this was also a leftover.
i later realized it's actually not possible to reliably decide if a config is using old format since all of these keys are optional....

---

*[mobly/config_parser.py, line 205 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJBha59azoS3y41EZ:-KcVTw4XX-r6GV8nnguy:b-mdqnxr) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L205)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Why do we need separate external and internal representations?
</blockquote></details>

The intention was so that we could change internal structure and names without affecting user-facing keys, or vice versa, which has proven to be useful, especially in this case....

---

*[mobly/config_parser.py, line 237 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJh3J9zJALRghZpJG:-KcVUB9r45VXnb4YinOU:b-yqs4yq) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L237)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Is this supposed to be private to the module?
If you don't want external callers, needs an underscore.
</blockquote></details>

No, it was intended for people to call so we have a single entry point in loading any config file in Mobly tests.
But today it is less important to have this.

---

*[mobly/keys.py, line 30 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJoUu9AEfz6u0BlrX:-KcVV3Pio5BZC8T-vpGl:b-xbuuga) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/keys.py#L30)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

based on the convention above, seems like this should be key_test_beds?
</blockquote></details>

The benefit of using these keys is so that we don't have to change all the references when we want to change keys. I think it's ok to just keep the vars' names.

---

*[mobly/test_runner.py, line 201 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVK7N0CY5P2Ka1ja2J:-KcVUeXetVHbXxojHt8e:b-2gw7m4) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/test_runner.py#L201)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Different newline conventions between this line and the line above?
</blockquote></details>

this is yapf's doing...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 8 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/google/mobly/85#-KcVLsTW4KC_CEGYcrGm:-KcVW5jV-WV5zniQSsQ4:bb6un1j) (no related file):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

android_device has some references to json configs in comments
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[README.md, line 213 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGMsh-jB4DfpfrUFL:-KcZp4Ey20NvhkiV3V9C:b-m39d8d) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/README.md#L213)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Does this work? Wikipedia says comments in yaml are # not //

---

*[setup.py, line 28 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGqMKvoR1czWx45KQ:-KcZp9KhiiBwHQ3-8MHp:b-5qxxi3) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/setup.py#L28)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Done
</blockquote></details>

f < m :)

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-KcZpBdOHiSa6Jw9T3EM:b-jd18tm) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

All the docs files were copied from readme... do we want to remove them for good?
We never got the time to make them pretty... :(
</blockquote></details>

We should remove whatever is the unsupported version of the documentation.

---

*[mobly/config_parser.py, line 205 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJBha59azoS3y41EZ:-KcZpXGL9hGBkRxQDTXa:b6ykbqt) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L205)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

The intention was so that we could change internal structure and names without affecting user-facing keys, or vice versa, which has proven to be useful, especially in this case....
</blockquote></details>

If it's part of an API it should probably be a real python object with attributes instead of another dict.

---

*[mobly/config_parser.py, line 237 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJh3J9zJALRghZpJG:-KcZpd-FYUBTjjaQEatU:b2p35sy) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L237)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

No, it was intended for people to call so we have a single entry point in loading any config file in Mobly tests.
But today it is less important to have this.
</blockquote></details>

So can we make it private?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[README.md, line 213 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGMsh-jB4DfpfrUFL:-Kc_nDzU6iRl1eX0zX1J:bb6un1j) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/README.md#L213)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Does this work? Wikipedia says comments in yaml are # not //
</blockquote></details>

Done

---

*[setup.py, line 28 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVGqMKvoR1czWx45KQ:-Kc_nHb-FVEkK_1lEZcn:b-qa4wj7) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/setup.py#L28)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

f < m :)
</blockquote></details>

whaaaaa?!

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-Kc_nImRizUEGcBXpgKs:b7n75ze) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

We should remove whatever is the unsupported version of the documentation.
</blockquote></details>

Removing these docs would mean removing the entire 'docs' dir?

---

*[mobly/config_parser.py, line 205 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJBha59azoS3y41EZ:-Kc_s8UbS4w3dUi6M9Bj:b-zfwihr) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L205)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

If it's part of an API it should probably be a real python object with attributes instead of another dict.
</blockquote></details>

Why was it a dict in the first place................ ʕノ•ᴥ•ʔノ ︵ ┻━┻

---

*[mobly/config_parser.py, line 237 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJh3J9zJALRghZpJG:-Kc_ndOzHJmtbvQXBXXE:bb6un1j) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L237)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

So can we make it private?
</blockquote></details>

Done

---

*[tests/mobly/test_runner_test.py, line 44 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVKt-ryjP7ogDePwQ1:-Kc_nvCTJh057LnoB329:b-hbb45z) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/tests/mobly/test_runner_test.py#L44)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Why is this config using internal keys? This used to look like a parsed external config
</blockquote></details>

A parsed config uses internal keys.
And this is a parsed config that is passed to TestRunner directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[mobly/config_parser.py, line 204 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVJMIgIe5gYuXc1zWc:-Kc_sQD89zCsN-xHz0UG:b-hv947o) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/config_parser.py#L204)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Since this object seems to have fixed keys and is part of an API, seems like it should be a real python object instead of a dict.
</blockquote></details>

yes...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Reviewed 6 of 7 files at r6.
Review status: 16 of 17 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-Kcu_9rSUQ8-8hdXRW05:bu31p1s) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Removing these docs would mean removing the entire 'docs' dir?
</blockquote></details>

We want to move these to wiki, right? When should we do that?

---

*[mobly/base_test.py, line 59 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_QGeX6osLTBFmEj1:-Kcu_QGeX6osLTBFmEj2:bnvlmva) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/base_test.py#L59)):*
> ```Python
>     TAG = None
> 
>     def __init__(self, configs):
> ```

Could we have a method description?
ie what type of thing is configs?

---

*[mobly/base_test.py, line 436 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_KDjtWv9arqZnvYH:-Kcu_KDknTMJJ7hoR9Rb:bx4u3b7) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/base_test.py#L436)):*
> ```Python
>         """
>         test_names = []
>         for name, _ in inspect.getmembers(self, inspect.ismethod):
> ```

What's with this change? Why is it related to yaml?

---

*[mobly/config_parser.py, line 267 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_qfHuw6FGdOUrArQ:-Kcu_qfHuw6FGdOUrArR:b-ue0ol9) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/config_parser.py#L267)):*
> ```Python
> 
>     def __str__(self):
>         return str(self.__dict__)
> ```

github is complaining that there's new newline at the end of the file

---

*[mobly/keys.py, line 36 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_zHEY8i41drcKZT3:-Kcu_zHEY8i41drcKZT4:bex99ta) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/keys.py#L36)):*
> ```Python
>     # Internal keys, used internally to detach the internal var names from
>     # keywords that define the config format.
>     ikey_user_param = "user_params"
> ```

What are the ikeys used for now?

---

*[mobly/test_runner.py, line 170 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuaHcvR9olIPGTvU9W:-KcuaHcvR9olIPGTvU9X:by97x8h) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L170)):*
> ```Python
> 
>     Attributes:
>         self.test_run_info: A dictionary containing the information needed by
> ```

Is this still a dict?

---

*[mobly/test_runner.py, line 176 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcua6w_Bhboh7uynM4w:-Kcua6w_Bhboh7uynM4x:bil0i7k) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L176)):*
> ```Python
>         self.test_configs: A TestRunConfig object with the information needed
>                            to execute a test run.
>         test_bed_name: The name of the test bed to execute the test on.
> ```

Inconsistent use of self in this comment

---

*[mobly/test_runner.py, line 194 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuaZ2G23jEdycQ_jM3:-KcuaZ2Ht_VPiyyG5R4r:bue7g4i) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L194)):*
> ```Python
>     """
> 
>     def __init__(self, test_configs, run_list):
> ```

Can you add a docstring? ie what kind of thing is test_configs?

---

*[mobly/test_runner.py, line 382 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuadgPi4CvQ0LE5Xbc:-KcuadgPi4CvQ0LE5Xbd:bdpppya) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L382)):*
> ```Python
>         into a dictionary to be passed to test classes.
>         """
>         self.test_run_info = copy.deepcopy(self.test_configs)
> ```

Food for thought : maybe we should implemenent the copy ability inside self.test_configs itself?

---

*[mobly/test_runner.py, line 385 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuapQaMqgFQCmFgkNg:-KcuapQbxM_V0JiiCesT:bu5lfea) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L385)):*
> ```Python
>         self.test_run_info.log_path = self.log_path
>         self.test_run_info.log = self.log
>         self.test_run_info.cli_args = getattr(self.test_configs, 'cli_args', None)
> ```

I thought everything in test_configs started out being set to None? Why the extra getattr?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 16 of 17 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-KczZOWO-vZlalqcMxwF:b-jbmqlv) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

We want to move these to wiki, right? When should we do that?
</blockquote></details>

That seems like a separate topic from this PR?
We could remove them after this PR.

---

*[mobly/base_test.py, line 59 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_QGeX6osLTBFmEj1:-Kcza9LwJXFwN6rcKM8E:bb6un1j) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/base_test.py#L59)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Could we have a method description?
ie what type of thing is configs?
</blockquote></details>

Done

---

*[mobly/base_test.py, line 436 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_KDjtWv9arqZnvYH:-KczaBQzWouSqnR_K3u-:b-ehqcle) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/base_test.py#L436)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

What's with this change? Why is it related to yaml?
</blockquote></details>

This is a bug fix. Before this, we'll try to execute any attribute whose name starts with `test_`, even if the attribute itself is a string.

---

*[mobly/config_parser.py, line 267 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_qfHuw6FGdOUrArQ:-KczaQjKpXl8MC0Cl1yo:bb6un1j) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/config_parser.py#L267)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

github is complaining that there's new newline at the end of the file
</blockquote></details>

Done

---

*[mobly/keys.py, line 36 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_zHEY8i41drcKZT3:-Kczb0iIal1AFlSz6Cbl:bxnd0vx) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/keys.py#L36)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

What are the ikeys used for now?
</blockquote></details>

Done.
I always forget to do my favorite part - deleting code >__<

---

*[mobly/test_runner.py, line 170 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuaHcvR9olIPGTvU9W:-KczbdTHjKKFZbyfZQNg:bb6un1j) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L170)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Is this still a dict?
</blockquote></details>

Done

---

*[mobly/test_runner.py, line 176 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcua6w_Bhboh7uynM4w:-KczbhMJZKushlKdYg9M:bb6un1j) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L176)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Inconsistent use of self in this comment
</blockquote></details>

Done

---

*[mobly/test_runner.py, line 194 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuaZ2G23jEdycQ_jM3:-Kczda7ETaE9yjz8LnL2:bb6un1j) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L194)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Can you add a docstring? ie what kind of thing is test_configs?
</blockquote></details>

Done

---

*[mobly/test_runner.py, line 382 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuadgPi4CvQ0LE5Xbc:-KczcnVU_QG_VO_-WYit:bfrp1ig) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L382)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Food for thought : maybe we should implemenent the copy ability inside self.test_configs itself?
</blockquote></details>

This is the only place there this happens right now, so maybe later.

---

*[mobly/test_runner.py, line 385 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuapQaMqgFQCmFgkNg:-KczbvoIOgtGgWtfmdH7:b-74z0ud) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L385)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

I thought everything in test_configs started out being set to None? Why the extra getattr?
</blockquote></details>

`cli_args` is actually not a predefined member of `test_configs`.
I want to remove the whole code path of it since it is not used, but not in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 12 of 17 files reviewed at latest revision, 12 unresolved discussions.

---

*[mobly/base_test.py, line 59 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_QGeX6osLTBFmEj1:-Kcze0iH_jdM1OCK9WUk:b-896fix) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/base_test.py#L59)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[mobly/base_test.py, line 436 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_KDjtWv9arqZnvYH:-Kcze1Ut88NJBelXUQsr:b-896fix) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/base_test.py#L436)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

This is a bug fix. Before this, we'll try to execute any attribute whose name starts with `test_`, even if the attribute itself is a string.
</blockquote></details>

Done.

---

*[mobly/keys.py, line 36 at r6](https://reviewable.io:443/reviews/google/mobly/85#-Kcu_zHEY8i41drcKZT3:-Kcze3nDM6P66_iooAdF:b-896fix) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/keys.py#L36)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Done.
I always forget to do my favorite part - deleting code >__<
</blockquote></details>

Done.

---

*[mobly/test_runner.py, line 201 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVK7N0CY5P2Ka1ja2J:-Kcze4ZcEJKUl7qYvO9U:b-896fix) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/mobly/test_runner.py#L201)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

this is yapf's doing...
</blockquote></details>

Done.

---

*[mobly/test_runner.py, line 194 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuaZ2G23jEdycQ_jM3:-Kcze5cAmfBtZ3ae-gpI:b-896fix) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L194)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Ping
Unit test flaked, pls ignore that...

adorokhine commented:
Reviewed 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-KdE6qH5IAbZZ0qA7GYI:b-z73jmy) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

That seems like a separate topic from this PR?
We could remove them after this PR.
</blockquote></details>

Could you rebase?

---

*[mobly/base_test.py, line 74 at r7](https://reviewable.io:443/reviews/google/mobly/85#-KdE72Tyet7OVJbDOCtE:-KdE72Tyet7OVJbDOCtF:bg57s4z) ([raw file](https://github.com/google/mobly/blob/5f9a7bf58bb7e56e0f7bf05f349eed25424a15f8/mobly/base_test.py#L74)):*
> ```Python
>             self.TAG = self.__class__.__name__
>         # Set params.
>         for name, value in configs.__dict__.items():
> ```

TestRunConfig has specific fields now, right?
Can you avoid this magic and just manually set the fields you need to expose as part of base_test?
Shouldn't need to look into the private __dict__ hash of regular python objects.

---

*[mobly/keys.py, line 43 at r7](https://reviewable.io:443/reviews/google/mobly/85#-KdE7ivO3clsQCJBljiU:-KdE7ivO3clsQCJBljiV:b-c4ua06) ([raw file](https://github.com/google/mobly/blob/5f9a7bf58bb7e56e0f7bf05f349eed25424a15f8/mobly/keys.py#L43)):*
> ```Python
> 
> 
> def value_to_value(ref_value, pattern):
> ```

No longer used

---

*[mobly/test_runner.py, line 382 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuadgPi4CvQ0LE5Xbc:-KdE8Fuw6DEAbI0TnVHr:byykyjp) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L382)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

This is the only place there this happens right now, so maybe later.
</blockquote></details>

What do you mean it's the only place this happens right now? There's 10 instances of this.
```
mobly/test_runner.py:            controller_config = copy.deepcopy(original_config)
mobly/test_runner.py:        self.test_run_info = copy.deepcopy(self.test_configs)
tests/mobly/base_test_test.py:        configs = copy.deepcopy(self.mock_test_cls_configs)
tests/mobly/base_test_test.py:        configs = copy.deepcopy(self.mock_test_cls_configs)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
What's wrong with the unit tests? Seems like they should be fixed instead of asking reviewers to ignore them?

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85#-:-KdE8hBhHTQNp8fWF9yn:b-3ncjqf)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 14 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 14 at r4](https://reviewable.io:443/reviews/google/mobly/85#-KcVH2VtWZbBzJnd7y0o:-Kdb8jY2lrtEkGKswaw8:bkxwonl) ([raw file](https://github.com/google/mobly/blob/e1a6bab863ff291c56473be70ff1c550b239d7fc/docs/tutorials/lesson1.md#L14)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Could you rebase?
</blockquote></details>

Rebase on what?

---

*[mobly/base_test.py, line 74 at r7](https://reviewable.io:443/reviews/google/mobly/85#-KdE72Tyet7OVJbDOCtE:-KdbCUB8WiZgTed_7CWO:b-896fix) ([raw file](https://github.com/google/mobly/blob/5f9a7bf58bb7e56e0f7bf05f349eed25424a15f8/mobly/base_test.py#L74)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

TestRunConfig has specific fields now, right?
Can you avoid this magic and just manually set the fields you need to expose as part of base_test?
Shouldn't need to look into the private __dict__ hash of regular python objects.
</blockquote></details>

Done.

---

*[mobly/keys.py, line 43 at r7](https://reviewable.io:443/reviews/google/mobly/85#-KdE7ivO3clsQCJBljiU:-Kdb8wZfcZ0OcMvSGW3r:b-896fix) ([raw file](https://github.com/google/mobly/blob/5f9a7bf58bb7e56e0f7bf05f349eed25424a15f8/mobly/keys.py#L43)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

No longer used
</blockquote></details>

Done.

---

*[mobly/test_runner.py, line 382 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuadgPi4CvQ0LE5Xbc:-Kdb9ImW5-NkiuiuXFTG:b-w1vlvh) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L382)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

What do you mean it's the only place this happens right now? There's 10 instances of this.
```
mobly/test_runner.py:            controller_config = copy.deepcopy(original_config)
mobly/test_runner.py:        self.test_run_info = copy.deepcopy(self.test_configs)
tests/mobly/base_test_test.py:        configs = copy.deepcopy(self.mock_test_cls_configs)
tests/mobly/base_test_test.py:        configs = copy.deepcopy(self.mock_test_cls_configs)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
tests/mobly/test_runner_test.py:        mock_test_config = copy.deepcopy(self.base_mock_test_config)
```
</blockquote></details>

`self.test_run_info = copy.deepcopy(self.test_configs)` is the only one in actual code.
The other one is test_runner is copying a dict, not the Config object.

All rest are in tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
The flaky unit test is already tracked by https://github.com/google/mobly/issues/97

---

Review status: 14 of 17 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85#-:-KdbSxa7r0_6eqa0YW4g:b-ehg6q5)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[mobly/base_test.py, line 79 at r8](https://reviewable.io:443/reviews/google/mobly/85#-KdhYU4bTKe6Av5dKKMt:-KdhYU4bTKe6Av5dKKMu:b8k80pr) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/mobly/base_test.py#L79)):*
> ```Python
>                                  controller modules.
>             log: a logger object. (deprecated)
>             cli_args: any cli args passed in. (deprecated)
> ```

Can we just remove this if it's deprecated and there are no users? same with log

---

*[mobly/keys.py, line 36 at r8](https://reviewable.io:443/reviews/google/mobly/85#-KdhYd5LdYsSwcoxlk76:-KdhYd5MyIOXey4EQhTM:b-l0ic7v) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/mobly/keys.py#L36)):*
> ```Python
> 
> 
> def get_name_by_value(value):
> ```

No longer used?

---

*[mobly/test_runner.py, line 382 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuadgPi4CvQ0LE5Xbc:-KdhYlwF-ILoQQCYCBSd:bpqp72n) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L382)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

`self.test_run_info = copy.deepcopy(self.test_configs)` is the only one in actual code.
The other one is test_runner is copying a dict, not the Config object.

All rest are in tests.
</blockquote></details>

Do tests not count as code? I'm not sure I understand the resistance to having a clone() method which is a one-liner that does a deep copy; what's the harm?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 12 at r8](https://reviewable.io:443/reviews/google/mobly/85#-KdhZA4WA-cyZISGbBuZ:-KdhZA4WA-cyZISGbBu_:b-ceopkf) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/docs/tutorials/lesson1.md#L12)):*
> ```Markdown
> device's screen. Create the following files:
> 
> **sample_config.yml**
> ```

This file has been deleted right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[mobly/base_test.py, line 79 at r8](https://reviewable.io:443/reviews/google/mobly/85#-KdhYU4bTKe6Av5dKKMt:-Kdhsypu8bUuw-Lbktbj:byk3t91) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/mobly/base_test.py#L79)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Can we just remove this if it's deprecated and there are no users? same with log
</blockquote></details>

lots of test are using `self.log`, so it can't be removed. There's an internal tracking bug to remove those usage.
removing cli_arg will be in a separate PR as it touches a lot of files.

---

*[mobly/keys.py, line 36 at r8](https://reviewable.io:443/reviews/google/mobly/85#-KdhYd5LdYsSwcoxlk76:-Ke29mBqwNLHJxXXNxgi:b-896fix) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/mobly/keys.py#L36)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

No longer used?
</blockquote></details>

Done.

---

*[mobly/test_runner.py, line 382 at r6](https://reviewable.io:443/reviews/google/mobly/85#-KcuadgPi4CvQ0LE5Xbc:-Ke2Ah-o3-jLPQpKOI7C:b-896fix) ([raw file](https://github.com/google/mobly/blob/f3105d601c4c91666cc3d117dc24c96a52c0a453/mobly/test_runner.py#L382)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Do tests not count as code? I'm not sure I understand the resistance to having a clone() method which is a one-liner that does a deep copy; what's the harm?
</blockquote></details>

Done.

---

*[docs/tutorials/lesson1.md, line 12 at r8](https://reviewable.io:443/reviews/google/mobly/85#-KdhZA4WA-cyZISGbBuZ:-KdhsrnEEbeJGh1KHB-w:bncvegj) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/docs/tutorials/lesson1.md#L12)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

This file has been deleted right?
</blockquote></details>

Rebased.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Reviewed 5 of 10 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mobly/config_parser.py, line 279 at r9](https://reviewable.io:443/reviews/google/mobly/85#-KeahFBaniAtJAemc_cJ:-KeahFBbHaTHK0nKkhXS:b-phha87) ([raw file](https://github.com/google/mobly/blob/df35af64697a954cafa501a02c131a908c408e2b/mobly/config_parser.py#L279)):*
> ```Python
>         """Returns a deep copy of the current config.
>         """
>         return copy.deepcopy(self)
> ```

Does this work? I tried to reproduce this in a canned example and got a crash:

t.py:
```
import foo
print(foo.copy())
```

foo.py:
```
import copy
def copy():
  return copy.deepcopy({})
```

```
$ python t.py 
Traceback (most recent call last):
  File "t.py", line 3, in <module>
    print(foo.copy())
  File "/tmp/foo.py", line 4, in copy
    return copy.deepcopy({})
AttributeError: 'function' object has no attribute 'deepcopy'
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mobly/config_parser.py, line 279 at r9](https://reviewable.io:443/reviews/google/mobly/85#-KeahFBaniAtJAemc_cJ:-Keaj9nae6lbHGiE2m5U:b-duonut) ([raw file](https://github.com/google/mobly/blob/df35af64697a954cafa501a02c131a908c408e2b/mobly/config_parser.py#L279)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Does this work? I tried to reproduce this in a canned example and got a crash:

t.py:
```
import foo
print(foo.copy())
```

foo.py:
```
import copy
def copy():
  return copy.deepcopy({})
```

```
$ python t.py 
Traceback (most recent call last):
  File "t.py", line 3, in <module>
    print(foo.copy())
  File "/tmp/foo.py", line 4, in copy
    return copy.deepcopy({})
AttributeError: 'function' object has no attribute 'deepcopy'
```
</blockquote></details>

It works because the copy here is under the scope of the class, whereas the copy in your example is a standalone function which causes name conflicts in console.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 1 at r8](https://reviewable.io:443/reviews/google/mobly/85#-Keaykai82PhZRx1l4w4:-KeaykajvVBk3buz8Mqa:b-yy72fb) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/docs/tutorials/lesson1.md#L1)):*
> ```Markdown
> ---
> ```

docs/tutorial.md is affected, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 1 at r8](https://reviewable.io:443/reviews/google/mobly/85#-Keaykai82PhZRx1l4w4:-Keb0WxQ6pHMcm1bM6I6:b9fqe3y) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/docs/tutorials/lesson1.md#L1)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

docs/tutorial.md is affected, right?
</blockquote></details>

Oh god, rebase erased all my changes for this part...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mobly/base_test.py, line 69 at r9](https://reviewable.io:443/reviews/google/mobly/85#-Keaz3B4LxQoKak-qaJ8:-Keaz3B4LxQoKak-qaJ9:bh26i6u) ([raw file](https://github.com/google/mobly/blob/df35af64697a954cafa501a02c131a908c408e2b/mobly/base_test.py#L69)):*
> ```Python
>         Args:
>             configs: A config_parser.TestRunConfig object.
>             log_path: string, specifies the root directory for all logs written
> ```

Everything other than configs aren't Args. Did you mean to put them into an Attributes section or something like that?

---

*[docs/tutorials/lesson1.md, line 1 at r8](https://reviewable.io:443/reviews/google/mobly/85#-Keaykai82PhZRx1l4w4:-Keb38xu9a68-tmb9pJx:b8ej8q5) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/docs/tutorials/lesson1.md#L1)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Oh god, rebase erased all my changes for this part...
</blockquote></details>

Luckily it's still in the old reviewable revisions. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 11 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/tutorials/lesson1.md, line 1 at r8](https://reviewable.io:443/reviews/google/mobly/85#-Keaykai82PhZRx1l4w4:-Keb9JXb2INHE4Y_ymGa:b-896fix) ([raw file](https://github.com/google/mobly/blob/4066dd9a6b3487f73e2ad24dd628f5bafb168e2f/docs/tutorials/lesson1.md#L1)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Luckily it's still in the old reviewable revisions. :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 11 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[mobly/base_test.py, line 69 at r9](https://reviewable.io:443/reviews/google/mobly/85#-Keaz3B4LxQoKak-qaJ8:-Keb9pNoZSAKnCelYo3k:b-896fix) ([raw file](https://github.com/google/mobly/blob/df35af64697a954cafa501a02c131a908c408e2b/mobly/base_test.py#L69)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Everything other than configs aren't Args. Did you mean to put them into an Attributes section or something like that?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Reviewed 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[mobly/base_test.py, line 64 at r10](https://reviewable.io:443/reviews/google/mobly/85#-KebP_yLRTiJkY2oWONj:-KebP_yN9MvVmuDTENyZ:b-1tbitf) ([raw file](https://github.com/google/mobly/blob/5d0f19e88d73b44fe71c06e7ac04230afc8493dd/mobly/base_test.py#L64)):*
> ```Python
>         register_controller: func, used by test classes to register
>                              controller modules.
>         log: a logger object. (deprecated)
> ```

Duplicate entry for 'log' in attributes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[mobly/base_test.py, line 64 at r10](https://reviewable.io:443/reviews/google/mobly/85#-KebP_yLRTiJkY2oWONj:-KebTdHNDOl6Js7aLKDV:b-896fix) ([raw file](https://github.com/google/mobly/blob/5d0f19e88d73b44fe71c06e7ac04230afc8493dd/mobly/base_test.py#L64)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Duplicate entry for 'log' in attributes
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[mobly/keys.py, line 30 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebYaxVuu7FBPMyFqav:-KebYaxWOq2bg2tiQ6Kh:b6lxip9) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/mobly/keys.py#L30)):*
> ```Python
>     key_testbed_name = "Name"
>     key_testbed_controllers = "Controllers"
>     key_testbed_test_params = "TestParams"
> ```

Should be using single quotes in new code, not double quotes (here and elsewhere)

---

*[mobly/test_runner.py, line 396 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebYMe4KfQc5RUF1a0l:-KebYMe55Sb43hBg1B0M:b-52sq2o) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/mobly/test_runner.py#L396)):*
> ```Python
>         self.test_run_info.log_path = self.log_path
>         self.test_run_info.log = self.log
>         self.test_run_info.cli_args = getattr(self.test_configs, 'cli_args', None)
> ```

Why do we have to go through getattr for this? cli_args is initialized to None right?
```
        # Below attributes are deprecated and kept for compatibility reasons
        # for now.
        self.log = None
        self.cli_args = None
```

---

*[tests/mobly/base_test_test.py, line 766 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebYwgqz2ffiINVU19Y:-KebYwgqz2ffiINVU19Z:b731hgc) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/tests/mobly/base_test_test.py#L766)):*
> ```Python
>         required = ["something"]
>         optional = ["something_else"]
>         configs = copy.deepcopy(self.mock_test_cls_configs)
> ```

Not using the new copy method? (here and elsewhere)

---

*[tests/mobly/test_runner_test.py, line 17 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebZ9D8rII2-uqo5nMm:-KebZ9D8rII2-uqo5nMn:br9o9r4) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/tests/mobly/test_runner_test.py#L17)):*
> ```Python
> # limitations under the License.
> 
> import copy
> ```

Is this import still used?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[mobly/keys.py, line 30 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebYaxVuu7FBPMyFqav:-KebbP_rzyU5vAPeHqsu:b-896fix) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/mobly/keys.py#L30)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Should be using single quotes in new code, not double quotes (here and elsewhere)
</blockquote></details>

Done.

---

*[mobly/test_runner.py, line 396 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebYMe4KfQc5RUF1a0l:-Kebal5DLhgmkmwbBCMa:b-thdjhb) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/mobly/test_runner.py#L396)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Why do we have to go through getattr for this? cli_args is initialized to None right?
```
        # Below attributes are deprecated and kept for compatibility reasons
        # for now.
        self.log = None
        self.cli_args = None
```
</blockquote></details>

Oh it didn't used to. Done

---

*[tests/mobly/base_test_test.py, line 766 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebYwgqz2ffiINVU19Y:-Kebb8bx37_Wk-YmqS2q:b-896fix) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/tests/mobly/base_test_test.py#L766)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Not using the new copy method? (here and elsewhere)
</blockquote></details>

Done.

---

*[tests/mobly/test_runner_test.py, line 17 at r11](https://reviewable.io:443/reviews/google/mobly/85#-KebZ9D8rII2-uqo5nMm:-KebbErlr2N96qtPMN9d:b-896fix) ([raw file](https://github.com/google/mobly/blob/709e0e5bc9c855956395eda0a34f1a8aaf7f75a3/tests/mobly/test_runner_test.py#L17)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Is this import still used?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/85#-:-KeeXEDMgaWBNG9d5vBe:bnfp4nl)*
<!-- Sent from Reviewable.io -->
